File: 2472-maximum-number-of-non-overlapping-palindrome-substrings/2472-maximum-number-of-non-overlapping-palindrome-substrings.cpp
class Solution {
public:
    int maxPalindromes(string s, int k) {
        int n=s.length();
        vector<vector<bool>> isPalindrome(n+1, vector<bool>(n+1, false));
        for(int L=1;L<=n;L++){
            for(int i=0;i+L<=n;i++){
                int j=i+L-1;
                if(i==j) isPalindrome[i][j]=true; // length 1
                else if(i+1==j) isPalindrome[i][j]=(s[i]==s[j]);
                else isPalindrome[i][j]=(s[i]==s[j] && isPalindrome[i+1][j-1]);
            }
        }
        vector<int> dp(n+1, 0);
        for(int L=k;L<=n;L++){
            int result=dp[L-1];
            int j=L-1;
            for(int i=0;j-i+1>=k;i++){
                if(isPalindrome[i][j]){
                    result=max(result,1+dp[i]);
                }
            }
            dp[L]=result;
        }
        return dp[n];
    }
};

File: 2472-maximum-number-of-non-overlapping-palindrome-substrings/2472-maximum-number-of-non-overlapping-palindrome-substrings.js
/**
 * @param {string} s
 * @param {number} k
 * @return {number}
 */
var maxPalindromes = function(s, k) {
    const n=s.length;
    const isPalindrome=Array(n+1).fill(false).map(()=> Array(n+1).fill(false));
    for(let L=1;L<=n;L++){
        for(let i=0;i+L<=n;i++){
            let j=i+L-1;
            if(i===j) isPalindrome[i][j]=true; // length 1
            else if(i+1===j) isPalindrome[i][j]=(s[i]===s[j]);
            else isPalindrome[i][j]=(s[i]===s[j] && isPalindrome[i+1][j-1]); 
        }
    }
    const dp=Array(n+1).fill(0);
    for(let L=k;L<=n;L++){
        let result=dp[L-1];
        let j=L-1;
        for(let i=0;j-i+1>=k;i++){
            if(isPalindrome[i][j]){
                result=Math.max(result,1+dp[i]);
            }
        }
        dp[L]=result;
    }
    return dp[n];
};